File: solutions/mpi/fortran/posix-io/separate-files.f90
program pario
  use mpi
  use, intrinsic :: iso_fortran_env, only : error_unit, output_unit
  implicit none

  integer, parameter :: datasize = 64, writer_id = 0
  integer :: rc, my_id, ntasks, localsize, i
  integer, dimension(:), allocatable :: localvector
  integer, dimension(datasize) :: fullvector

  call mpi_init(rc)
  call mpi_comm_size(mpi_comm_world, ntasks, rc)
  call mpi_comm_rank(mpi_comm_world, my_id, rc)

  if (ntasks > 64) then
    write(error_unit, *) 'Maximum number of tasks is 64!'
    call mpi_abort(MPI_COMM_WORLD, -1, rc)
  end if

  if (mod(datasize, ntasks) /= 0) then
    write(error_unit,*) 'Datasize (64) should be divisible by number of tasks'
    call mpi_abort(MPI_COMM_WORLD, -1, rc)
  end if

  localsize = datasize / ntasks
  allocate(localvector(localsize))

  localvector = [(i + my_id * localsize, i=1,localsize)]

  call many_writers()

  deallocate(localvector)
  call mpi_finalize(rc)

contains

  subroutine single_writer()
    implicit none

    call mpi_gather(localvector, localsize, mpi_integer, fullvector, &
         & localsize, mpi_integer, writer_id, mpi_comm_world, rc)
    if (my_id == writer_id) then
      open(10, file='singlewriter.dat', status='replace', form='unformatted', &
           & access='stream')
      write(10, pos=1) fullvector
      close (10)
      write(output_unit,'(A,I0,A)') 'Wrote ', size(fullvector), &
           & ' elements to file singlewriter.dat'
    end if
  end subroutine single_writer

  subroutine many_writers()
    implicit none
    character(len=85) :: filename

    write(filename, '(A,I0,A)') 'manywriters-', my_id, '.dat'

    open(my_id+10, file=filename, status='replace', form='unformatted', &
         & access='stream')
    write(my_id+10, pos=1) localvector
    close (my_id+10)
    write(output_unit,'(A,I0,A,A)') 'Wrote ', size(localvector), &
         & ' elements to file ', filename
  end subroutine many_writers

end program pario

File: mpi/posix-io/c/solution/separate-files.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <mpi.h>

#define DATASIZE   64
#define WRITER_ID   0

void single_writer(int, int *, int);
void many_writers(int, int *, int);


int main(int argc, char *argv[])
{
    int my_id, ntasks, i, localsize;
    int *localvector;

    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &ntasks);
    MPI_Comm_rank(MPI_COMM_WORLD, &my_id);

    if (ntasks > 64) {
        fprintf(stderr, "Datasize (64) should be divisible by number "
                "of tasks.\n");
        MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
    }

    if (DATASIZE % ntasks != 0) {
        fprintf(stderr, "Datasize (64) should be divisible by number "
                "of tasks.\n");
        MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
    }

    localsize = DATASIZE / ntasks;
    localvector = (int *) malloc(localsize * sizeof(int));

    for (i = 0; i < localsize; i++) {
        localvector[i] = i + 1 + localsize * my_id;
    }

    many_writers(my_id, localvector, localsize);

    free(localvector);

    MPI_Finalize();
    return 0;
}

void single_writer(int my_id, int *localvector, int localsize)
{
    FILE *fp;
    int *fullvector;

    fullvector = (int *) malloc(DATASIZE * sizeof(int));

    MPI_Gather(localvector, localsize, MPI_INT, fullvector, localsize,
               MPI_INT, WRITER_ID, MPI_COMM_WORLD);

    if (my_id == WRITER_ID) {
        if ((fp = fopen("singlewriter.dat", "wb")) == NULL) {
            fprintf(stderr, "Error: %d (%s)\n", errno, strerror(errno));
            MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
        } else {
            fwrite(fullvector, sizeof(int), DATASIZE, fp);
            fclose(fp);
            printf("Wrote %d elements to file singlewriter.dat\n", DATASIZE);
        }
    }

    free(fullvector);
}

void many_writers(int my_id, int *localvector, int localsize)
{
    FILE *fp;
    char filename[64];

    sprintf(filename, "manywriters-%d.dat", my_id);

    if ((fp = fopen(filename, "wb")) == NULL) {
        fprintf(stderr, "Error: %d (%s)\n", errno, strerror(errno));
        MPI_Abort(MPI_COMM_WORLD, EXIT_FAILURE);
    } else {
        fwrite(localvector, sizeof(int), localsize, fp);
        fclose(fp);
        printf("Wrote %d elements to file manywriters-%d.dat\n", localsize,
                my_id);
    }
}

